### PR TITLE
fromPromise carries forward old value from pending observable

### LINKED ISSUE #305: fromPromise sets pending value to undefined if oldPromise is also pending
I found a race in code using `fromPromise`. When rapidly sending new requests in quick succession, created with `fromPromise`, the `case` method will temporarily return `undefined` until the latest response returns. Based on the documentation, I would expect it to always return stale data and never return `undefined`.

## Example scenario
```
@observable
private dataResponse = undefined;

private loadData() {
    this.dataResponse = fromPromise(fetch(...), this.dataResponse);
}

getDataForReactUI() {
    return this.dataResponse?.case({
        fulfilled: (result) => result,
        pending: (stale) => stale,
    });
}
```

The race appears when the fetch response is slow, and the user triggers another load quickly. The sequence is:
 1. `loadData()` called for the first time, and `this.dataResponse` is initialized.
 2. Response returns and `this.data` resolves. User sees the returned data.
 3. User clicks a refresh button.
 4. `loadData()` is called a second time; `fromPromise()` is called with the resolved promise as the second argument.
 5. User clicks  the refresh button again before the previous request returns.
 6. `loadData()` is called a third time; `fromPromise()` is called with the *pending* promise as the second argument.
 
### What happens:
User sees data disappear for a few seconds, before seeing the new data appear.
### What I expect 
User sees stale data until new data is available.

## Root cause
The cause seems to be this code inside `fromPromise()`:
https://github.com/mobxjs/mobx-utils/blob/v5.6.1/src/from-promise.ts#L185
```
    const oldData =
        oldPromise && (oldPromise as any).state === FULFILLED
            ? (oldPromise as any).value
            : undefined
```

In short, if the `oldPromise` is still PENDING, `fromPromise()` will use `undefined` even if the `oldPromise` has a value for its pending state.

### PR DESCRIPTION
**Pull request by laszlopandy**
Fixes #305

### CODE REVIEW

File: src/from-promise.ts [hunk 1]
```diff
@@ -181,10 +181,9 @@ export function fromPromise<T>(
 
     promise.isPromiseBasedObservable = true
     promise.case = caseImpl
+    const oldState = oldPromise ? (oldPromise as any).state : undefined
     const oldData =
-        oldPromise && (oldPromise as any).state === FULFILLED
-            ? (oldPromise as any).value
-            : undefined
+        oldState === FULFILLED || oldState === PENDING ? (oldPromise as any).value : undefined
```
Comment (urugator) [inline src/from-promise.ts:186]: So ... `oldState !== REJECTED` ? :smile:

Comment (laszlopandy) [reply src/from-promise.ts:186]: Yes, but I thought enumerating both would make the reason more explicit.

Comment (urugator) [reply src/from-promise.ts:186]: Sure it doesn't really matter, feel free to keep it that way. I was just thinking about it the other way around ... if we carry over the previous promise value, why don't we carry over the errored value as well ... so from this perspective the REJECTED state acts as exception.

**Comment by urugator** [comment]
Could you perhaps update changelog? Not sure how changelog in this package is supposed to be managed... :)
Otherwise LGTM, good catch, thank you!

**Comment by laszlopandy** [comment]
> Could you perhaps update changelog? Not sure how changelog in this package is supposed to be managed... :) Otherwise LGTM, good catch, thank you!

Done!

**Pull request merged**